### PR TITLE
Feature/local exception translator

### PR DESCRIPTION
**Pull request by jesse-sony**
## Description

Allow exception translators to be optionally registered local to a module instead of applying globally across all pybind11 modules.

This is useful in a situation where there are more than one compiled modules (e.g., clients for different micro-services) that share a lot of the same core code on the c++ side, including exceptions. Shared classes can be handled via the py::module_local() tag on the classes, but there is currently no way to deal with exception remapping.

So in code like:
```
    py::register_exception_translator([](std::exception_ptr p) {
      try {
        if (p) {
          std::rethrow_exception(p);
        }
      }
      catch (const ::vasg::exceptions::VasgException& e) {
        auto err = py::cast(e);
        auto errType = err.get_type().ptr();
        PyErr_SetObject(errType, err.ptr());
      }
```
The exception translator is global, so which module's exception you get back is determined by import order.

As an example, if you did something like
```
import tempo # compiled lib 

tempo.resolveKey(badPath)
```
you get back an exception from the tempo module.

However, if you were to do
```
import rhapsody # compiled lib
import tempo # compiled lib

tempo.resolveKey(badPath)
```
you'd get back an exception from the rhapsody module. Since those are different types in python (but the same type in c++), it means catch statements which are expecting the exception to come from the same module as the call are failing.

## Suggested changelog entry:

```rst

Allow exception translators to be optionally registered local to a module instead of applying globally across all pybind11 modules.
Use 
register_local_exception_translator(ExceptionTranslator&& translator)
instead of 
register_exception_translator(ExceptionTranslator&& translator)
to keep your exception remapping code local to the module.
```

### Why does this not require an Internals Version bump?
The internals struct in pybind11 is shared between all the modules. Any changes to that definitely requires a lot of careful thought about whether backwards compatibility is being broken. The local_internals that are added here should, by design, be segregated to a single module. Whether a module has local internals or not should not impact any other modules, because the only things accessing the local internals is the module that contains them.

### CODE REVIEW

**Review by rhaschke — APPROVED** [review]
I think this is an important feature and the implementation looks reasonable to me.

File: include/pybind11/detail/internals.h
Comment (YannickJadoul) [inline include/pybind11/detail/internals.h:322]: This should probably not be called `locals` anymore, since it's not just the local variables anymore?

File: include/pybind11/detail/internals.h
Comment (YannickJadoul) [inline include/pybind11/detail/internals.h:327]: Not sure how other parts of pybind11 do it, but you do have a duplicate use of `registered_local_types_cpp` here (field as well as the function name). Can you check what conventions other parts of `internals.h` follow?

Comment (YannickJadoul) [reply include/pybind11/detail/internals.h:327]: Well, I overcame my laziness and looked up myself. Can we go for `get_registered_local_types_cpp` as function name, for example? I don't think this should break ABI things.

File: include/pybind11/detail/internals.h
Comment (YannickJadoul) [inline include/pybind11/detail/internals.h:315]: Can/should we make this `module_local_internals`?

Comment (YannickJadoul) [reply include/pybind11/detail/internals.h:315]: (same for `get_module_local_internals`, then, if we do)

File: include/pybind11/pybind11.h
Comment (YannickJadoul) [inline include/pybind11/pybind11.h:780]: Can you update the comment above? I.e., the "give each registered exception translator a chance to translate it to a Python exception in reverse order of registration."

File: include/pybind11/pybind11.h
Comment (YannickJadoul) [inline include/pybind11/pybind11.h:1956]: > "This function will be applied"

"These will be tried ..." or so? "applied" seems weird

**Review by YannickJadoul — COMMENTED** [review]
I general, I'm happy with this (but I was also involved in the Gitter discussion, so I'm not entirely well-placed to judge this).
This is also one of the things @wjakob should probably have a look at? Does it make sense to always check module-local exception translators before the global, shared ones?

We're still lacking changes to the actual documentation, by the way. A note about the existence of module-local exception/exception translators, and stating the order in which they are resolved is important.

Also, what about the auxiliary function `register_exception`? Should it get an extra argument, or somehow use the `py::module_local()` tag, or an extra template argument, or a sibling function `register_module_local_exception`, or ... ?

File: include/pybind11/detail/internals.h
Comment (jesse-sony) [reply include/pybind11/detail/internals.h:322]: it was all stuff that was local to the module, so seemed appropriate. I'm happy to rename it though. Since it never escapes the context of the function, not sure how much it really matters

File: include/pybind11/detail/internals.h
Comment (jesse-sony) [reply include/pybind11/detail/internals.h:327]: would you prefer me to rename this function or just replace usages of it with `get_module_internals().registered_local_types_cpp;` so it works more like get_internals?

Comment (jesse-sony) [reply include/pybind11/detail/internals.h:327]: or would that break ABI? I don't think this function should be ever used in user code

File: include/pybind11/detail/internals.h
Comment (jesse-sony) [reply include/pybind11/detail/internals.h:315]: I'd do either `local_internals` or `module_internals` to contrast with the normal `internals`. I don't see a reason to give it all the names. but if that is what people want, I'll do it

File: include/pybind11/detail/internals.h
Comment (YannickJadoul) [reply include/pybind11/detail/internals.h:322]: Doesn't really matter that much, indeed. Just something that caught my eye, since I thought the old name was linked to Python's `locals()`; maybe it isn't, though.

File: include/pybind11/detail/internals.h
Comment (YannickJadoul) [reply include/pybind11/detail/internals.h:327]: Matching `get_internals()` closer might be nice, indeed, yes, if you ask me :-)
And no, this shouldn't break user code (since it's in the `detail` namespace), and ... yeah, no, the whole point is that this isn't shared between modules, so there's no shared ABI :-)

File: include/pybind11/detail/internals.h
Comment (YannickJadoul) [reply include/pybind11/detail/internals.h:315]: I was thinking of "module local" being a concept, but it's true it also wasn't done for `registered_local_types_cpp`. Maybe matching `local` is more important than `module`, though?

Let's see what others say.

File: include/pybind11/pybind11.h
Comment (jesse-sony) [reply include/pybind11/pybind11.h:784]: done

File: include/pybind11/pybind11.h
Comment (jesse-sony) [reply include/pybind11/pybind11.h]: verbiage has been adjusted

Comment (jesse-sony) [reply include/pybind11/pybind11.h]: You can tell I've been doing a lot of functional programming lately, because I'm out here applying functions to things

File: include/pybind11/detail/internals.h
Comment (jesse-sony) [reply include/pybind11/detail/internals.h:322]: I added a `register_local_exception` function, as well.  Someone who is better at c++ template wrangling than me could probably remove the amount of copy-pasta between register_exception and register_local_exception.

File: include/pybind11/detail/internals.h
Comment (jesse-sony) [reply include/pybind11/detail/internals.h]: I have updated this to be called `get_local_internals()`, deleted `get_registered_local_types_cpp()` completely, and changed the usage to be more like `get_internals()`. Do we want to leave the double `local` in the names or match the member names to the `internals` member names?
currently:
```
get_local_internals().registered_local_types_cpp;
```
proposed change to remove the local in the member name
```
get_local_internals().registered_types_cpp;
```

File: include/pybind11/detail/internals.h
Comment (jesse-sony) [reply include/pybind11/detail/internals.h]: currently went with `local_internals` and `get_local_internals`. obviously easy to change depending on what everyone wants

**Comment by jesse-sony** [comment]
Documentation has also been updated

File: include/pybind11/pybind11.h
Comment (YannickJadoul) [reply include/pybind11/pybind11.h]: Hah :-D Thanks!

**Review by YannickJadoul — APPROVED** [review]
Looks good to me. If this gets one or two more reviews, I think it can be merged :-)

File: include/pybind11/pybind11.h
Comment (jesse-sony) [reply include/pybind11/pybind11.h:794]: The sort of thing that `chain` from python is perfect for but doesn't seem to be std lib version of it

File: docs/advanced/exceptions.rst
Comment (jesse-sony) [reply docs/advanced/exceptions.rst:72]: done

File: docs/advanced/exceptions.rst
Comment (jesse-sony) [reply docs/advanced/exceptions.rst]: removed

File: docs/advanced/exceptions.rst
Comment (jesse-sony) [reply docs/advanced/exceptions.rst]: Redid that whole section to be less duplicative.

File: docs/advanced/exceptions.rst
Comment (YannickJadoul) [reply docs/advanced/exceptions.rst]: Thanks :-)

File: include/pybind11/cast.h
Comment (jesse-sony) [reply include/pybind11/cast.h:174]: okay. My preference is to remove the second local too. Hopefully other people agree

File: tests/test_exceptions.cpp
Comment (jesse-sony) [reply tests/test_exceptions.cpp:9]: done

**Comment by rwgk** [comment]
Thanks @Skylion007 for pointing out this PR, I missed it before. I didn't look at all details, but I see others have already. My thinking: @jesse-sony if you rebase and resolve merge conflicts, I'll run the refreshed version through the Google-global testing (includes testing a large number of open source packages) asap, if that checks out I'll merge.

**Comment by jesse-sony** [comment]
@rwgk Branch has been rebased and pushed. Should be in a good state to merge

**Comment by rwgk** [comment]
Thanks! My testing will take a day or so.

**Review by rwgk — CHANGES_REQUESTED** [review]
The Google-global testing came back green, feature-wise this PR looks great to me, but the code duplication is a concern. I think it's important to go the extra mile for best practices straightaway. Usually such things don't get fixed later but rather tend to get duplicated even more (slippery slope).

File: include/pybind11/detail/internals.h
Comment (jesse-sony) [reply include/pybind11/detail/internals.h:324]: My thought process on this (and I'm definitely not an expert in all the intricacies of pybind11 ABI compatibility) is that the internals struct in pybind11 is shared between all the modules. Any changes to that definitely requires a lot of careful thought about whether backwards compatibility is being broken. The `local_internals` that I added here should, by design, be segregated to a single module. Whether a module has local internals or not should not impact any other modules, because the only things accessing the local internals is the module that contains them.
If this makes sense to you as well, I'll happily add a comment to the end of this PR explaining that thought process

File: docs/advanced/exceptions.rst
Comment (jesse-sony) [reply docs/advanced/exceptions.rst:230]: whoops

File: include/pybind11/cast.h
Comment (jesse-sony) [reply include/pybind11/cast.h]: @rwgk any thoughts on the naming in the `local_internals` struct?
```
struct local_internals {
  type_map<type_info *> registered_local_types_cpp;
  std::forward_list<void (*) (std::exception_ptr)> registered_local_exception_translators;
};
```
vs
```
struct local_internals {
  type_map<type_info *> registered_types_cpp;
  std::forward_list<void (*) (std::exception_ptr)> registered_exception_translators;
};
```

File: docs/advanced/exceptions.rst
Comment (jesse-sony) [reply docs/advanced/exceptions.rst]: removed

File: docs/advanced/exceptions.rst
Comment (jesse-sony) [reply docs/advanced/exceptions.rst]: added

File: include/pybind11/pybind11.h
Comment (jesse-sony) [reply include/pybind11/pybind11.h]: fixed

File: include/pybind11/pybind11.h
Comment (jesse-sony) [reply include/pybind11/pybind11.h]: done.

File: include/pybind11/pybind11.h
Comment (jesse-sony) [reply include/pybind11/pybind11.h]: Okay, I did this. I did not change any of the type information, because I'm pretty sure that would PYBIND11_INTERNALS_VERSION ABI breaking change (so internals still has
```
std::forward_list<ExceptionTranslator> registered_exception_translators;
```
not
```
std::forward_list<const ExceptionTranslator> registered_exception_translators;
```
as it would in your PR.

But I removed all the templating from the register functions. I also turned `ExceptionTranslator` into a `using` alias, which makes the typing a little bit nicer to read.

File: include/pybind11/pybind11.h
Comment (jesse-sony) [reply include/pybind11/pybind11.h]: this duplication has been removed and made into a helper. Let me know if you don't like something about my implementation.

**Comment by rwgk** [comment]
Hi @jesse-sony, today I only got a chance to glance through your comments really quickly, not even the code. The comments I saw look good, thanks!. I'll come back here asap.

File: include/pybind11/detail/internals.h
Comment (rwgk) [reply include/pybind11/detail/internals.h:328]: Make total sense to me, but I don't consider myself an authority on this, too. Adding this to the PR description is to expose it, in case someone is able to poke a hole into the rationale, we want to know asap.

File: include/pybind11/pybind11.h
Comment (rwgk) [reply include/pybind11/pybind11.h]: Yes, perfect, thanks. I did not mean to suggest adding the const. My PR was just to record that I thought about it carefully and then discarded the idea because I think it's too disruptive for too little gain.

File: include/pybind11/detail/internals.h
Comment (rwgk) [inline include/pybind11/detail/internals.h:322]: Answering the question you asked in a different comment:
I'd prefer the 2nd alternative you pointed out there, with `local` **not** repeated inside the struct.
But I'd approve either.

**Review by rwgk — APPROVED** [review]
Looks awesome, thanks!

I'm still struggling a bit navigating this review system. If I overlooked a question you asked, please tag me again. (I tried to tie all my responses to the Review.)

I'm holding off merging until you make you final decision about the local_internals naming. Just let me know.

File: include/pybind11/detail/internals.h
Comment (jesse-sony) [reply include/pybind11/detail/internals.h]: Names have been changed, since everyone seems to prefer them without the `local`

**Comment by jesse-sony** [comment]
@rwgk Cleaned up the naming, updated the PR description with a note about why it does not break ABI compatability with the internals struct, and also add some notes in the internals.h file about the same ABI stuff. Happy to make other changes if there is more to be done, but I think I've hit everything mentioned so far.

**Comment by rwgk** [comment]
The one failing CI job is a flake:
```
Run jwlawson/actions-setup-cmake@v1.9
Error: getaddrinfo ENOTFOUND api.github.com
```
Safe to ignore.

**Comment by rwgk** [comment]
Thanks @jesse-sony for the change, @rhaschke, @YannickJadoul for the reviews, and @Skylion007 for pointing out this PR as one we should look at again. I'll merge this now.

**Pull request merged**
